src/core_modules/capture-core/components/WidgetProfile/DataEntry/hooks/useLifecycle.js

```diff
@@ -0,0 +1,123 @@
+// @flow
+import { useState, useEffect, useMemo } from 'react';
+import { useDispatch, useSelector } from 'react-redux';
+
+import type {
+    OrgUnit,
+    TrackedEntityAttributes,
+    OptionSets,
+    ProgramRulesContainer,
+} from 'capture-core-utils/rulesEngine';
+import { useProgramRules, useConstants } from './index';
+import { getOpenDataEntryActions } from '../actions';
+import { buildRulesContainer } from '../ProgramRules';
+import { buildFormFoundation } from '../FormFoundation';
+
+export const useLifecycle = ({
+    programAPI,
+    orgUnitId,
+    mergedAttributes,
+    dataEntryId,
+    itemId,
+}: {
+    programAPI: any,
+    orgUnitId: string,
+    mergedAttributes: Array<any>,
+    dataEntryId: string,
+    itemId: string,
+}) => {
+    // TODO: Getting the entire state object is bad and this needs to be refactored.
+    // The problem is the helper methods that take the entire state object.
+    // Refactor the helper methods (getCurrentClientValues, getCurrentClientMainData in rules/actionsCreator) to be more explicit with the arguments.
+    const state = useSelector(stateArg => stateArg);
+
+    const { programRules, loading: loadingProgramRules } = useProgramRules(programAPI.id);
+    const { constants } = useConstants();
+    const dispatch = useDispatch();
+    const [rulesContainer, setRulesContainer] = useState<ProgramRulesContainer>({});
+    const [formFoundation, setFormFoundation] = useState<any>({});
+
+    const orgUnit: OrgUnit = useMemo(() => ({ id: orgUnitId, name: '' }), [orgUnitId]);
+    const trackedEntityAttributes: TrackedEntityAttributes = useMemo(
+        () =>
+            programAPI?.programTrackedEntityAttributes?.reduce(
+                (acc, currentValue) => ({
+                    ...acc,
+                    [currentValue.trackedEntityAttribute.id]: currentValue.trackedEntityAttribute,
+                }),
+                {},
+            ),
+        [programAPI],
+    );
+    const optionSets: OptionSets = useMemo(
```

Comment:
We should also include the optionSets from the dataElements in the program stages here.